profile2/tool/src/java/org/sakaiproject/profile2/tool/entityprovider/ProfileEntityProvider.java

```diff
@@ -469,23 +473,76 @@ public Object getNamePronunciation(OutputStream out, EntityView view, Map<String
 		MimeTypeByteArray mtba = profileLogic.getUserNamePronunciation(uuid);
 		if(mtba != null && mtba.getBytes() != null) {
 			final byte[] bytes = mtba.getBytes();
+			int fileSize = bytes.length;
 			//check for binary
-			if(bytes != null && bytes.length > 0) {
+			if(bytes != null && fileSize > 0) {
 				try {
-					// Set mime type here before header is sent
-					requestGetter.getResponse().setContentType("audio/ogg");
-					out.write(bytes);
-					ActionReturn actionReturn = new ActionReturn(StandardCharsets.UTF_8.name(), "audio/ogg", out);
-
-					Map<String,String> headers = new HashMap<>();
-					headers.put("Expires", "0");
-					headers.put("Cache-Control","no-cache, no-store, must-revalidate");
-					headers.put("Pragma", "no-cache");
-					actionReturn.setHeaders(headers);
+					requestGetter.getResponse().reset();
+					requestGetter.getResponse().setHeader("Expires", "0");
+					requestGetter.getResponse().setHeader("Pragma", "no-cache");
+					requestGetter.getResponse().setHeader("Cache-Control", "no-cache, no-store, must-revalidate");
 
+					// Set mime type here before header is sent
+					requestGetter.getResponse().setContentType(mtba.getMimeType());
+					//Set the content disposition
+					requestGetter.getResponse().setHeader("Content-Disposition", String.format("attachment; filename=\"%s.wav\"; filename*=UTF-8''%s.wav", uuid, uuid));
+
+					//iOS and MacOS always send a range request for the file, we should return the requested part of the recording.
+					String range = requestGetter.getRequest().getHeader("range");
```

Comment:
Instead of including our own implementation for supporting http range we could just use springs HttpRange class? Its part of spring-web.

Reply:
I needed to borrow some Apple devices (iMac and iPad) to resolve this problem and I can't borrow them anymore, I can't work on this anymore and my client is reviewing the solution right now, it actually took me a lot of effort. 

If anyone has an Apple device and can work on this piece of source, would be wonderful.